Framework/src/Checker.cxx

```diff
@@ -70,120 +120,137 @@ Checker::~Checker()
                << diagnostic;
     throw;
   }
-  startFirstObject = system_clock::time_point::min();
-  timer.reset(1000000); // 10 s.
+
 }
 
 void Checker::run(framework::ProcessingContext& ctx)
 {
-  mLogger << "Receiving " << ctx.inputs().size() << " MonitorObjects" << AliceO2::InfoLogger::InfoLogger::endm;
+  mLogger << mCheckerNames.front() <<" Receiving " << ctx.inputs().size() << " MonitorObjects" << AliceO2::InfoLogger::InfoLogger::endm;
 
   // Save time of first object
   if (startFirstObject == std::chrono::system_clock::time_point::min()) {
     startFirstObject = system_clock::now();
   }
 
-  std::shared_ptr<TObjArray> moArray{ framework::DataRefUtils::as<TObjArray>(*ctx.inputs().begin()) };
-  moArray->SetOwner(false);
-  auto checkedMoArray = std::make_unique<TObjArray>();
-  checkedMoArray->SetOwner();
-
-  for (const auto& to : *moArray) {
-    std::shared_ptr<MonitorObject> mo{ dynamic_cast<MonitorObject*>(to) };
-    moArray->RemoveFirst();
-    if (mo) {
-      check(mo);
-      store(mo);
-      mTotalNumberHistosReceived++;
-      checkedMoArray->Add(new MonitorObject(*mo));
-    } else {
-      mLogger << "the mo is null" << AliceO2::InfoLogger::InfoLogger::endm;
+  for (const auto& input: mInputs){
+    auto dataRef = ctx.inputs().get(input.binding.c_str());
+    if(dataRef.header != nullptr && dataRef.payload != nullptr){
+      auto moArray = ctx.inputs().get<TObjArray*>(input.binding.c_str());
+      for (const auto& to : *moArray) {
+        std::shared_ptr<MonitorObject> mo{ dynamic_cast<MonitorObject*>(to) };
+
+        if (mo) {
+          update(mo);
+          mTotalNumberHistosReceived++;
+        } else {
+          mLogger << "The mo is null" << AliceO2::InfoLogger::InfoLogger::endm;
+        }
+      }
     }
   }
 
-  send(checkedMoArray, ctx.outputs());
+  // Check if compliant with policy
+  if (mPolicy->isReady()){
+    auto qualityVector = check(mMonitorObjects);
+    store(qualityVector);
+    //send(checkedMoArray, ctx.outputs());
+  }
 
   // monitoring
   endLastObject = system_clock::now();
-
-  // if 10 seconds elapsed publish stats
   if (timer.isTimeout()) {
     timer.reset(1000000); // 10 s.
     mCollector->send({ mTotalNumberHistosReceived, "objects" }, o2::monitoring::DerivedMetricMode::RATE);
   }
 }
 
-o2::header::DataDescription Checker::createCheckerDataDescription(const std::string taskName)
-{
-  if (taskName.empty()) {
-    BOOST_THROW_EXCEPTION(FatalException() << errinfo_details("Empty taskName for checker's data description"));
-  }
-  o2::header::DataDescription description;
-  description.runtimeInit(std::string(taskName.substr(0, o2::header::DataDescription::size - 4) + "-chk").c_str());
-  return description;
+void Checker::update(std::shared_ptr<MonitorObject> mo){
+  mLogger << mCheckerNames.front() << " - moMap key: " << mo->getTaskName() << AliceO2::InfoLogger::InfoLogger::endm;
+  mMonitorObjects[mo->getTaskName()] = mo;
```

Comment:
Wrong assumption, task can deliver different MOs under the same InputSpec.

Reply:
I would change the `MonitorObject::getName()` -> `::getObjectName()` and `MonitorObject::getName() { return getTaskName() + "/"+ getObjectName(); }` with the assumption that getName will return unique name across the QC.
What do you think @knopers8 ?

Reply:
I have nothing against it, though maybe `getFullName` would be better than `getName`. What I am afraid of is that, the MonitorObject interface is also used by QCG (QC GUI) to show MOs, so that change could break something. Barth is more aware of that part of QC, so I would wait until he's back.

Reply:
Using here `getName()` should be probably enough to have it working now, but indeed, there is a risk that two QC Tasks may produce an object of an identical name.